### ISSUE TITLE
V8 Causing Shiny Crash

### ISSUE THREAD

**Issue by GrayAlex49** [jeroen/V8#119]
Rendering plots that call V8 in shiny are either not working or completely crashing shiny. 

Verified by the shiny team with reproducible example - [rstudio/shiny#3289](https://github.com/rstudio/shiny/issues/3289#issuecomment-778305691)

When I tried to make a reproducable example of this issue I was having (segfaults as seen in the issue above) I created a minimal example that doesn't crash but still will not render in shiny although it does in Rstudio ([malcolmbarrett/ggdag#55](https://github.com/malcolmbarrett/ggdag/issues/55)). Interestingly it happens on windows and docker but not mac os for R > 4.0.0. I had to roll back to R 3.6.3 to avoid this error.

**Comment by JanMarvin** [comment]
What does "happens on docker" mean? I cannot reproduce either crash with R 4.0.5 and V8 9.2.172 on Arch. It might be caused by a bug in the V8 version used. Can you reproduce it with the [static V8 build](https://github.com/jeroen/V8#linux-static-libv8) too?

**Comment by GrayAlex49** [comment]
My production image is what is crashing at the same place in docker as it is on windows (trying to draw a ggdag). I was working on recreating the docker example but I'm having trouble even getting a working docker image for a minimal example.

Both of the following give me the error:
```
v8_test  |   unable to load shared object '/usr/local/lib/R/site-library/V8/libs/V8.so':
v8_test  |   libnode.so.64: cannot open shared object file: No such file or directory
```

```
FROM rocker/shiny-verse:latest

RUN apt-get update && apt-get install -y \
  libv8-dev

RUN install2.r -e\
    V8 
```

```
FROM rocker/shiny-verse:latest

RUN R -e "Sys.setenv(DOWNLOAD_STATIC_LIBV8 = 1); \
    install.packages('V8')"
```

**Comment by jeroen** [comment]
@GrayAlex49 your last error is a completely different issue: `libnode.so.64: cannot open shared object file: No such file or directory`. This just means that you need to `apt-get install libnode-dev` on your image.

You can probably also solve it by switching to another docker images, I think `rocker/shiny-verse:latest` is no longer updated (it has R 3.6.3, right?). I very much recommend using a docker image based on at least ubuntu 20.04 and R 4.0 or newer, see https://github.com/rocker-org/rocker-versioned2#version-stable-rocker-images-for-r--400

**Comment by GrayAlex49** [comment]
@jeroen I am aware this is a completely separate issue, I use V8 I'm just trying to get to the bottom of a bug that is catching me upstream. 

Adding `apt-get install libnode-dev` does not solve the libnode.so.64 issue. As stated on the install guide for Ubuntu "On the latest systems, libv8-dev is actually an alias for libnode-dev so they are the same:" I tried installing both, no change.

[rocker/shiny-verse:latest](https://hub.docker.com/r/rocker/shiny-verse/tags?page=1&ordering=last_updated) is updated, my session info on that test gives me R version 4.0.5  Running under: Ubuntu 20.04.2 LTS

**Comment by jeroen** [comment]
Can you show some output of what you see then please. On ubuntu 20, `apt-get install libv8-dev` should pull in [libnode64](https://packages.ubuntu.com/focal/libnode64) which contains `libnode.so.64`

**Comment by GrayAlex49** [comment]
In the build log (after update and upgrade complete)
```
#4 42.03 Preparing to unpack .../libuv1_1.34.2-1ubuntu1.1_amd64.deb ...
#4 42.04 Unpacking libuv1:amd64 (1.34.2-1ubuntu1.1) ...
#4 42.08 Selecting previously unselected package libuv1-dev:amd64.
#4 42.08 Preparing to unpack .../libuv1-dev_1.34.2-1ubuntu1.1_amd64.deb ...
#4 42.09 Unpacking libuv1-dev:amd64 (1.34.2-1ubuntu1.1) ...
#4 42.13 Selecting previously unselected package libc-ares2:amd64.
#4 42.14 Preparing to unpack .../libc-ares2_1.15.0-1build1_amd64.deb ...
#4 42.14 Unpacking libc-ares2:amd64 (1.15.0-1build1) ...
#4 42.19 Selecting previously unselected package libnode64:amd64.
#4 42.19 Preparing to unpack .../libnode64_10.19.0~dfsg-3ubuntu1_amd64.deb ...
#4 42.20 Unpacking libnode64:amd64 (10.19.0~dfsg-3ubuntu1) ...
#4 42.68 Selecting previously unselected package libnode-dev:amd64.
#4 42.68 Preparing to unpack .../libnode-dev_10.19.0~dfsg-3ubuntu1_amd64.deb ...
#4 42.69 Unpacking libnode-dev:amd64 (10.19.0~dfsg-3ubuntu1) ...
#4 42.75 Setting up libc-ares2:amd64 (1.15.0-1build1) ...
#4 42.77 Setting up libuv1:amd64 (1.34.2-1ubuntu1.1) ...
#4 42.78 Setting up libnode64:amd64 (10.19.0~dfsg-3ubuntu1) ...
#4 42.80 Setting up libuv1-dev:amd64 (1.34.2-1ubuntu1.1) ...
#4 42.82 Setting up libnode-dev:amd64 (10.19.0~dfsg-3ubuntu1) ...
#4 42.83 Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
#4 DONE 42.9s
```
And installing V8
```
#5 [3/3] RUN install2.r -e    V8
#5 sha256:6d3693d240aec7386471115befe15a6d86d23bb6c86f3c8f760588decc90dc1a
#5 1.676 trying URL 'https://packagemanager.rstudio.com/all/__linux__/focal/latest/src/contrib/V8_3.4.2.tar.gz'
#5 2.248 Content type 'binary/octet-stream' length 1357375 bytes (1.3 MB)
#5 2.249 ==================================================
#5 2.303 downloaded 1.3 MB
#5 2.303
#5 2.608 * installing *binary* package ‘V8’ ...
#5 2.613 * DONE (V8)
#5 2.618
#5 2.618 The downloaded source packages are in
#5 2.618        ‘/tmp/downloaded_packages’
#5 DONE 2.6s
```
Same error 
```
v8_test  | Error in dyn.load(file, DLLpath = DLLpath, ...) :
v8_test  |   unable to load shared object '/usr/local/lib/R/site-library/V8/libs/V8.so':
v8_test  |   libnode.so.64: cannot open shared object file: No such file or directory
v8_test  | Calls: runApp ... getNamespace -> loadNamespace -> library.dynam -> dyn.load
v8_test  | Execution halted
```
Built it again with the static V8 build 
```
#5 3.791 * installing *source* package ‘V8’ ...
#5 3.796 ** package ‘V8’ successfully unpacked and MD5 sums checked
#5 3.796 ** using staged installation
#5 3.815 Found C++14 compiler: g++
#5 4.016 > curl::curl_download("http://jeroen.github.io/V8/get-v8-linux.sh","get-v8-linux.sh")
#5 4.164 >
#5 4.164 >
#5 4.307 > curl::curl_download('http://jeroen.github.io/V8/v8-8.3.110.13-linux.tar.gz','libv8.tar.gz',quiet=FALSE)
 [100%] Downloaded 10522980 bytes...
#5 4.894 >
#5 4.894 >
#5 5.123 Using CXXCPP=g++ -std=gnu++14 -E
#5 5.123 Using PKG_CFLAGS=-I/tmp/RtmpQ0NOct/R.INSTALL2b5c327ca1/V8/v8/include
#5 5.123 Using PKG_LIBS=-L/tmp/RtmpQ0NOct/R.INSTALL2b5c327ca1/V8/v8/lib -lv8_monolith
#5 5.201 Running feature test for pointer compression...
#5 6.535 Pointer compression not needed
#5 6.549 ** libs
#5 6.552 rm -f V8.so RcppExports.o bindings.o
#5 6.554 g++ -std=gnu++14 -I"/usr/local/lib/R/include" -DNDEBUG -I/tmp/RtmpQ0NOct/R.INSTALL2b5c327ca1/V8/v8/include -DV8_ENABLE_CHECKS -I'/usr/local/lib/R/site-library/Rcpp/include' -I/usr/local/include  -fvisibility=hidden -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c RcppExports.cpp -o RcppExports.o
#5 9.941 g++ -std=gnu++14 -I"/usr/local/lib/R/include" -DNDEBUG -I/tmp/RtmpQ0NOct/R.INSTALL2b5c327ca1/V8/v8/include -DV8_ENABLE_CHECKS -I'/usr/local/lib/R/site-library/Rcpp/include' -I/usr/local/include  -fvisibility=hidden -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c bindings.cpp -o bindings.o
#5 13.40 g++ -std=gnu++14 -shared -L/usr/local/lib/R/lib -L/usr/local/lib -o V8.so RcppExports.o bindings.o -L/tmp/RtmpQ0NOct/R.INSTALL2b5c327ca1/V8/v8/lib -lv8_monolith -L/usr/local/lib/R/lib -lR
#5 14.36 installing to /usr/local/lib/R/site-library/00LOCK-V8/00new/V8/libs
#5 14.38 ** R
#5 14.38 ** inst
#5 14.39 ** byte-compile and prepare package for lazy loading
#5 14.79 ** help
#5 14.80 *** installing help indices
#5 14.81 ** building package indices
#5 14.97 ** installing vignettes
#5 14.97 ** testing if installed package can be loaded from temporary location
#5 15.21 ** checking absolute paths in shared objects and dynamic libraries
#5 15.22 ** testing if installed package can be loaded from final location
#5 15.46 ** testing if installed package keeps a record of temporary installation path
#5 15.46 * DONE (V8)
```
Same error as above

**Comment by jeroen** [comment]
How is it the same error? The last run seems like it works?

**Comment by GrayAlex49** [comment]
Sorry same error when running the app
```
v8_test  | Error in dyn.load(file, DLLpath = DLLpath, ...) :
v8_test  |   unable to load shared object '/usr/local/lib/R/site-library/V8/libs/V8.so':
v8_test  |   libnode.so.64: cannot open shared object file: No such file or directory
v8_test  | Calls: runApp ... getNamespace -> loadNamespace -> library.dynam -> dyn.load
v8_test  | Execution halted
```

**Comment by jeroen** [comment]
You must be doing something wrong in your path and having two copies of the v8 package intalled. It is impossible that the version that the last static build from above is asking for `libnode.so.64` because it hasn't been built with that.

**Comment by GrayAlex49** [comment]
I deleted the image and pruned the build layers between tests, I dont know what else to do. According to the docs, it should work either way but neither are working for me.

**Comment by GrayAlex49** [comment]
Okay I dont know what was going on with my docker builds, it seems to be sorted out. To get back to the original problem (I'm happy to have this issue closed and open a new one if it would be cleaner)

Dockerfile
```
FROM rocker/r-ver:4.0.0

RUN /rocker_scripts/install_shiny_server.sh

RUN R -e "Sys.setenv(DOWNLOAD_STATIC_LIBV8 = 1); \
    install.packages('V8', repos = 'https://cran.r-project.org')"

EXPOSE 3838

CMD ["/init"]
```
App
```
library(shiny)
V8::v8()     


ui <- fluidPage( plotOutput("myplot"), textOutput("info"))
server <- function(input, output) {
  output$myplot <- renderPlot({
    ctx <- V8::v8()   
    ctx$eval('"Hello World"')
    plot(1)
  })

  output$info <- renderPrint({
    sessionInfo()
  })
}
shinyApp(ui = ui, server = server, options = list(host='0.0.0.0', port=8000, launch.browser = FALSE))
```
Cause the error 
```
v8_test  | Warning: Error in context_eval: <string conversion failed>
v8_test  |   173: <Anonymous>
```
Which is the same error I see running it on windows.

By switching to the same R build but the one they built on 18.04 instead of 20.04 (```FROM rocker/r-ver:4.0.0-ubuntu18.04```) there is no error and the plot renders fine.

**Comment by jeroen** [comment]
Can you show for both machines the `V8::engine_info()` that is used in the shiny app?

**Comment by GrayAlex49** [comment]
Both are "8.3.110.13". That caused me to second guess myself so I wanted to double check with session info.
```
$version [1] "8.3.110.13"
R version 4.0.0 (2020-04-24) Platform: x86_64-pc-linux-gnu (64-bit) Running under: Ubuntu 20.04.2 LTS
```
```
$version [1] "8.3.110.13"
R version 4.0.0 (2020-04-24) Platform: x86_64-pc-linux-gnu (64-bit) Running under: Ubuntu 18.04.5 LTS 
```

Windows where shiny crashes $version [1] "6.2.414.50"

**Comment by JanMarvin** [comment]
Added it all into a single Dockerfile. Just for reference it works in R and is triggered only in the interaction with shiny. Maybe something OS specific on their end?

```
#syntax=docker/dockerfile:1.2
FROM rocker/r-ver:4.0.0

RUN /rocker_scripts/install_shiny_server.sh

RUN R -e "Sys.setenv(DOWNLOAD_STATIC_LIBV8 = 1); \
    install.packages('V8', repos = 'https://cran.r-project.org')"

RUN echo '\n\
library(shiny) \n\
V8::v8() \n\
\n\
\n\
ui <- fluidPage( plotOutput("myplot"), textOutput("info"))\n\
  server <- function(input, output) {\n\
             output$myplot <- renderPlot({\n\
               ctx <- V8::v8()\n\
               ctx$eval('"'"'"Hello World"'"'"')\n\
               plot(1)\n\
             })\n\
\n\
             output$info <- renderPrint({\n\
               sessionInfo()\n\
             })\n\
           }\n\
           shinyApp(ui = ui, server = server, options = list(host="0.0.0.0", port=8000, launch.browser = FALSE))' >> /srv/shiny-server/app.R

EXPOSE 3838

CMD ["/init"]
```

**Comment by GrayAlex49** [comment]
@JanMarvin Could be, but that would imply there is a separate bug in windows that also only affected by that same interaction. With this plot it plots normally in rstudio on windows but completely crashes shiny when running on windows. Also, I dont know how to create an example only using V8 but when plotting with certain packages that use V8 on windows it is possible to get the string conversion failed error in shiny (no crash).

**Comment by JanMarvin** [comment]
I expect it to be some kind of stacktrace. I can reproduce the error if I call `V8` more than once. E.g. if I run `library(V8)` or `V8::v8()`. Similar things happen if I call `pkgload::load_all(path = "/path/to/V8")` twice in a single session. This crashes my entire session (spotted this running `devtools::test()`).

**Comment by JanMarvin** [comment]
I've messed around with this issue [here (not very clean)](https://github.com/JanMarvin/V8/tree/V8_running). I have a solution that would solve this issue, but it breaks (almost everything) else. Basically, if you move the creation of `v8::Isolate` from `start_v8_isolate()` to `make_context()` you can run the code causing this issue (this creates isolates with the context (aaaand everything crashes if you create multiple isolated contexts and try to access them)).

Therefore I assume, that it's somehow a threading problem. Maybe `renderPrint()` forks the thread in an unusual way (could not reproduce the issue with `parallel::mclapply`). The issue appears when the context tries to access an isolate which is apparently not in the same thread. If I create the isolates per context (as described), this is avoided. (Forking would explain why the issue appears not on Windows.)

For reference:
In my branch I've tried to avoid creating new isolates and calling `v8::V8::Initalize()` in `start_v8_isolate()`.  Right now I read the process id and compare it to the previous process id and run it only if the are not identical. This solves the issue @jeroen opened [here](https://github.com/r-lib/pkgload/issues/96). But `V8` will still bail once `pkgload::load_all()` has been called and if it was not re-initialized (haven't found a way to check if this is required). V8 is quite sensitive to all the `dyn.load()` and threading stuff.

```r
pkgdown::build_site(devel=F) # works with my branch
pkgload::load_all() # works
ctx <- V8::v8() # crashes
```

**Comment by jeroen** [comment]
I think these issues should all be fixed. 

The error that sometimes appeared `<string conversion failed>` was caused by hitting the stack limit. This should be fixed in V8 3.6.0.

The other error that Gary mentioned above `libnode.so.64: cannot open shared object file: No such file or directory` is caused by some mix up of nodejs versions on your machine. I don't know how you did this, but we now default to using a stack build of libv8 on x64, so we no longer depend on nodejs.
